docs/proposals/build-inputs-overhaul.md

```diff
@@ -0,0 +1,309 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: build-inputs-overhaul
+authors:
+  - "@adambkaplan"
+  - "@otaviof"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional 
+replaces:
+  - "/docs/proposals/remote-artifacts.md"
+---
+
+# Build Inputs Overhaul
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+TBD
+
+## Summary
+
+Overhaul the Build API to separate build inputs from build execution and outputs.
+This overhaul will provide a path to supporting the following:
+
+1. Multiple build sources
+2. Build volume mounts
+
+## Motivation
+
+Building a container image can require more than one type of source as code/input.
+Builds may also rely on caches and other bits of information that are needed as input, but should not be present in the resulting container image.
+The current build API does not satisfy these two needs, and the current structure is not amenable to future adaptation.
+
+This proposal seeks to reorganize the Build API specification into three groups:
+
+1. Input
+2. Build (execution)
+3. Output
+
+Do so will allow Shipwright Builds to support multiple source inputs as well as mounts in builds.
+
+### Goals
+
+- Allow developers to assemble applications from multiple sources
+- Allow developers to mount volumes within the build
+
+### Non-Goals
+
+- Provide specific implementations for a given volume mount.
+- Provide additional source types beyond the ones supported today.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented. Include as much detail as
+possible so that people can understand the "how" of the system. The goal here is to make this feel
+real for users without getting bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+The Build API will serve as an abstraction layer for Tekton workspaces that are created in the associated TaskRun.
+The new `input` section of the build spec will declare how the Tekton workspaces for the BuildRun are configured.
+These workspaces will allow information to be shared across steps in the BuildRun.
+
+#### New API
+
+```yaml
+spec:
+  input:
+    sources:
```

Comment:
I am thinking if there is a duplicate between the `input` and `sources`, do we have any other properties under `input` besides `sources`?

If no, I think sources may be duplicate, then how about :
```
  input:
    - name: git
      type: Git
      git:
        url: https://github.com/shipwright-io/build.git
        ref: main
        credentials:
          name: git-creds
    - name: jars
      type: HTTP
      http:
        url: https://my-artifacts.corp/bin/base-lib.jar
      destination: bin/jars/ # Destination must be relative
...
```